thriftrw/spec/service.pyx

```diff
@@ -174,6 +175,39 @@ cdef class FunctionResultSpec(UnionTypeSpec):
             name, result_specs, allow_empty=(return_spec is None)
         )
 
+    cpdef object read_from(FunctionResultSpec self, ProtocolReader reader):
+        reader.read_struct_begin()
+
+        cdef dict kwargs = {}
+        cdef object val
+        cdef FieldSpec spec
+        cdef FieldHeader header
+
+        header = reader.read_field_begin()
+
+        # We use -1 to signify struct end due to cython constraints.
+        while header.type != -1:
+            spec = self._index.get((header.id, header.type), None)
+
+            if spec is None:
+                raise UnknownExceptionError(
+                    (
+                        '"%s" received an unrecognized exception. '
+                        'Make sure your Thrift IDL is up-to-date with '
+                        'what the remote host is using.'
+                    ) % self.name,
+                )
+
+            val = spec.spec.read_from(reader) or spec.default_value
```

Comment:
If the field was present, you can't not read it. `read_from` must return a
value and it must be used. Defaults are applied automatically by `__init__` so
you don't need to do this.

Besides, FieldResultSpec can't have defaults anyway; it's a union.